FEA add GLM Newton-LSMR Solver on top of Newton-Cholesky

lorentzenchr opened the pull request:
#### Reference Issues/PRs
This is a follow-up on #23314.

#### What does this implement/fix? Explain your changes.
It uses the machinery of #23314 and implements a further Newton solver based on LSMR for GLMs. It uses the IRLS formulation of the Newton step, which can be read as a normal equation of a least squares problem. This least squares problem is then solved via LSMR with a careful choice of the stopping criterion (`atol`).

This solver is good for dense and sparse input, only uses matrix-vector products, can stop early when achieving the convergence criteria, can handle thin and wide data, can deal with singular (collinear) X, but may fall short for very ill-conditioned X.

#### Any other comments?
I have never seen the combination of IRLS with LSMR (or LSQR). Momentarily, this is my favorite solver :wink:

As promised, LSMR seems to be able to stop a little earlier and therefore is a little faster than LSQR.

lorentzenchr committed:
FEA add NewtonSolver, CholeskyNewtonSolver and QRCholeskyNewtonSolver

lorentzenchr committed:
Merge branch 'main' into glm_newton_cholesky

lorentzenchr committed:
ENH better singular hessian special solve

lorentzenchr committed:
CLN fix some typos found by reviewer

lorentzenchr committed:
TST assert ConvergenceWarning is raised

lorentzenchr committed:
MNT add BaseCholeskyNewtonSolver

ogrisel committed:
WIP colinear design in GLMs

ogrisel committed:
Merge branch 'glm_newton_cholesky' into glm-newton-colinear

lorentzenchr committed:
FIX _solve_singular

lorentzenchr committed:
FIX solver for singular cases in GLMs (#5)

lorentzenchr committed:
ENH add LSMR Newton solver

lorentzenchr committed:
ENH add LSMR Newton solver

ogrisel commented:
Interesting! Since the lsmr sover is the fastest but stops at a higher loss, maybe the stopping criterion are not directly comparable.

Could you please run benchmarks with a grid of tolerance values to check if one solver is Pareto optimal (uniformly dominates the others on this tasks)?

https://github.com/scikit-learn/scikit-learn/pull/23314#pullrequestreview-968769894

Also, you might want to contribute both those solvers and the Poisson regression task to the benchopt project:

https://benchopt.github.io/

lorentzenchr committed:
ENH improve value of atol in LSMR

lorentzenchr committed:
ENH improve value of atol in LSMR

ogrisel commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -20,13 +24,746 @@
+class LSMRNewtonSolver(NewtonSolver):
+    """LSMR based Newton solver.
+
+    The inner solver uses LSMR after the Newton update is cast into the IRLS
```
Could you please expand those acronyms and point to a standard reference for LSMR?

As far as I understand:

- LSMR stands for Least Squares with MINRES (as an alternative to Least Squares with QR) and is introduced in https://web.stanford.edu/group/SOL/software/lsmr/LSMR-SISC-2011.pdf
- IRLS: Iteratively Reweighted Least Squares

ogrisel commented:
> can deal with singular (collinear) X, but may fall short for very ill-conditioned X

Not sure what's this means. It can handle collinear data as long as `n_samples >> n_uniquely_informative_features`?

ogrisel commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -20,13 +24,746 @@
+        n_features = X.shape[1]
+        # We use a 1-d array instead of a 2-d diagonal array to save memory. We also
+        # simply omit to add a zero at the end for the intercept.
+        self.sqrt_P = np.full(
```
Maybe this would be renamed to `self.sqrt_P_diag` to be explicit that the `(n_features, n_features)` `P` matrix is not needed.

lorentzenchr replied:
I agree that this would be a better name. But it would cause several new line breaks and make code less readable. This is already the case with `self.l2_reg_strength` which is way to long for what it's worth, IMHO.

ogrisel replied:
Personnally I prefer the extra line breaks :)

But this is minor so do as you wish and feel free to mark this thread resolved.

ogrisel commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -20,13 +24,746 @@
+        # Estimated Frobenius norm of A and norm of residual r for tolerance of
+        # next iteration.
+        self.A_norm = result[5]
+        self.r_norm = result[3]
```
We might want to issue a warning (and suggest increasing the regularization) if `result[1] == 7` (`istop`) which means that `lsmr` has not converged with the default maxiter` (which should be `n_features` by construction in our case), possibly because of ill-conditioning.

Instead of warning might even want to experiment with doing more iterations of `lsmr` when we detect this situation according to the docstring of `lsmr`, see: https://github.com/scipy/scipy/blob/main/scipy/sparse/linalg/_isolve/lsmr.py

and then only warn if we do not converge with `maxiter=2 * n_features` in total for instance (e.g. running `lsmr` twice in a row if the first call resulted in `istop == 7`).

Note that issuing warning when the inner solver failed to converge while still proceeding with the outer loop might just result in a wall of warnings and a badly converged model. Maybe we could also choose to just early interrupt the outer solver loop with a `ConvergenceWarning` in such a case. This is related to: https://github.com/lorentzenchr/scikit-learn/pull/6

ogrisel commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -20,13 +24,746 @@
+        n_features = X.shape[1]
+        # This duplicates a bit of code from LinearModelLoss.gradient.
+        weights, _ = self.linear_loss.weight_intercept(self.coef)
+        self.gradient = np.empty_like(self.coef, dtype=self.coef.dtype)
```
we could pre-allocate this buffer in the `setup` step.

ogrisel replied:
Actually the call to `self.linear_loss.loss_gradient` in the `line_search` method also allocates its own gradient vector and override this attribute. So as long as we do not have a way to pass a pre-allocated buffer to `self.linear_loss.loss_gradient` there is no point.

The performance impact is probably negligible anyways.

lorentzenchr committed:
FIX false unpacking in

lorentzenchr committed:
Merge branch 'glm_newton_lsmr' of https://github.com/lorentzenchr/scikit-learn into glm_newton_lsmr

lorentzenchr committed:
DOC less acronyms and reference

lorentzenchr committed:
Merge branch 'main' into glm_newton_cholesky

lorentzenchr committed:
Merge branch 'glm_newton_cholesky' into glm_newton_lsmr

lorentzenchr committed:
TST add tests for unpenalized GLMs

lorentzenchr committed:
TST fix solutions of glm_dataset

lorentzenchr committed:
ENH add SVDFallbackSolver

lorentzenchr committed:
CLN remove SVDFallbackSolver

lorentzenchr committed:
ENH use gradient step for singular hessians

lorentzenchr committed:
ENH print iteration number in warnings

lorentzenchr committed:
TST improve test_linalg_warning_with_newton_solver

lorentzenchr committed:
Merge branch 'glm_newton_cholesky' into glm_newton_lsmr

lorentzenchr committed:
MNT clean up after merge

lorentzenchr committed:
CLN LinAlgWarning fron scipy.linalg

lorentzenchr committed:
Merge branch 'glm_newton_cholesky' into glm_newton_lsmr

lorentzenchr committed:
ENH improve tiny hessian case

lorentzenchr committed:
ENH more robust tol in LSMR

lorentzenchr committed:
ENH edge cases in LSMR

lorentzenchr committed:
ENH more robust hessian

lorentzenchr committed:
ENH increase maxls for lbfgs to make it more robust

lorentzenchr committed:
ENH add hessian_warning for too many negative hessian values

lorentzenchr committed:
CLN some warning messages

lorentzenchr committed:
ENH add lbfgs_step

lorentzenchr committed:
ENH use lbfgs_step for hessian_warning

lorentzenchr committed:
TST make them pass

lorentzenchr committed:
TST tweek rtol for lbfgs

lorentzenchr committed:
TST add rigoros test for GLMs

lorentzenchr committed:
TST improve test_warm_start

lorentzenchr committed:
ENH improve lbfgs options for better convergence

lorentzenchr committed:
CLN fix test_warm_start

lorentzenchr committed:
TST fix assert singular values in datasets

lorentzenchr committed:
CLN address most review comments

lorentzenchr committed:
ENH enable more vebosity levels for lbfgs

lorentzenchr committed:
DOC add whatsnew

lorentzenchr committed:
Merge branch 'main' into glm_tests

lorentzenchr committed:
CLN remove xfail and clean a bit

lorentzenchr committed:
CLN docstring about minimum norm

ogrisel committed:
More informative repr for the glm_dataset fixture cases

ogrisel committed:
Forgot to run black

lorentzenchr committed:
CLN remove unnecessary filterwarnings

lorentzenchr committed:
CLN address review comments

lorentzenchr committed:
Merge branch 'glm_tests' of https://github.com/lorentzenchr/scikit-learn into glm_tests

ogrisel committed:
Trigger [all random seeds] on the following tests:
test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

lorentzenchr committed:
CLN add comment for lbfgs ftol=64 * machine precision

lorentzenchr committed:
Merge branch 'glm_tests' of https://github.com/lorentzenchr/scikit-learn into glm_tests

lorentzenchr committed:
CLN XXX code comment

lorentzenchr committed:
Trigger [all random seeds] on the following tests:

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

lorentzenchr committed:
CLN link issue and remove code snippet in comment

lorentzenchr committed:
Trigger [all random seeds] on the following tests:

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

lorentzenchr committed:
CLN add catch_warnings

lorentzenchr committed:
Trigger [all random seeds] on the following tests:

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

lorentzenchr committed:
Merge branch 'main' into glm_tests

lorentzenchr committed:
Trigger [all random seeds] on the following tests:

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

lorentzenchr committed:
[all random seeds]

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

lorentzenchr committed:
Trigger with -Werror [all random seeds]

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

lorentzenchr committed:
ENH increase maxls to 50

lorentzenchr committed:
[all random seeds]

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

lorentzenchr committed:
Revert "Trigger with -Werror [all random seeds]"

This reverts commit 99f4cf99ca41b4ad2bdad537ad60f936970e3a88.

lorentzenchr committed:
Merge branch 'glm_tests' into glm_newton_cholesky

lorentzenchr committed:
TST add catch_warnings to filterwarnings

lorentzenchr committed:
TST adapt tests for newton solvers

lorentzenchr committed:
CLN cleaner gradient step with gradient_times_newton

lorentzenchr committed:
DOC add whatsnew

lorentzenchr committed:
ENH always use lbfgs as fallback

lorentzenchr committed:
TST adapt rtol

lorentzenchr committed:
TST fix test_linalg_warning_with_newton_solver

lorentzenchr committed:
CLN address some review comments

ogrisel committed:
Improve tests related to convergence warning on collinear data

ogrisel committed:
overfit -> fit

ogrisel committed:
Typo in comment

ogrisel committed:
Merge branch 'main' into glm_newton_cholesky

ogrisel committed:
Apply suggestions from code review

ogrisel committed:
Merge remote-tracking branch 'origin/main' into glm_newton_cholesky

lorentzenchr committed:
ENH fallback_lbfgs_solve
- Do not use lbfgs steps, fall back complete to lbfgs

lorentzenchr committed:
ENH adapt rtol

ogrisel committed:
Merge branch 'glm_newton_cholesky' into test-glm-collinear-data

ogrisel committed:
Improve test_linalg_warning_with_newton_solver

ogrisel committed:
Better comments

ogrisel committed:
Merge pull request #8 from ogrisel/test-glm-collinear-data

Stricter test for glms with newton-cholesky solver on collinear data

ogrisel committed:
Fixed Hessian casing and improved warning messages

ogrisel committed:
[all random seeds]

test_linalg_warning_with_newton_solver

ogrisel committed:
Ignore ConvergenceWarnings for now if convergence is good

lorentzenchr committed:
CLN remove counting of warnings

lorentzenchr committed:
ENH fall back to lbfgs if line search did not converge

lorentzenchr committed:
DOC better comment on performance bottleneck

lorentzenchr committed:
Merge branch 'glm_newton_cholesky' into glm_newton_lsmr

lorentzenchr commented:
### Notes on multi-class logistic / multinomial regression
Is this possible with newton-lsmr?
Consider 3 classes. We concatenate the coefficient/weight of each class, i.e. coef = (coeffs_class_1, coeffs_class_2, coeffs_class_3). The Hessian becomes
```
(X'    ) (W11 W12 W13) (X  )
(  X'  ) (W12 W22 W23) ( X )
(    X') (W13 W23 W33) (  X)
```
Each submatrix Wij is diagonal of dimension n_samples * n_samples, Wii=p_i(1-p_i) and Wij=-pi * p_j with predicted probabilities p_i. For LSMR to work, we need to find a square root of W, say V with V'V=W. One could achieve this via a cholesky decomposition, but W is very sparse and very large.
See Chapter 3.2 of https://www.jstatsoft.org/article/view/v075i03.

lorentzenchr commented:
#### LDL of multinomial covariance matrix
The square root free LDL decomposition, given in https://doi.org/10.1111/j.2517-6161.1992.tb01875.x, reads:
```
Mathematical sum convention, starting from 1.
p = n-dimentional vector of probabilities summing to 1.
C = diag(p) - p'p, meaning C_ij = p_i - p_i p_j

LDL (square root free Cholesky) decomposition of C = L' D L with lower triangular L and diagonal D.
q_0 = 1
q_i = 1 - sum(p_k, k=1..i)

L_ii = 1
for i>j:
L_ij = -p_i / q_i

D_ii = p_i * q_i / q_{i-1}

If the p_i sum to 1, then q_n = 0 and d_n = 0.
```

ogrisel committed:
Merge branch 'main' into glm_newton_lsmr

lorentzenchr commented:
Adding multiclass/multinomial would be super cool, but I guess this PR would explode. I'd rather refer it to a later PR.

lorentzenchr committed:
Merge branch 'main' into glm_newton_lsmr

lorentzenchr committed:
Merge branch 'glm_newton_lsmr' of https://github.com/lorentzenchr/scikit-learn into glm_newton_lsmr

lorentzenchr committed:
FIX merge conflict

lorentzenchr committed:
CLN fix typo

lorentzenchr committed:
CLN fix typos

lorentzenchr committed:
Merge branch 'main' into glm_newton_lsmr

lorentzenchr commented:
I will open a new PR to add LSMR. I'll keep the branch as it is important for https://github.com/lorentzenchr/notebooks/blob/master/bench_glm_newton_solvers_cholesky_lsmr.ipynb.
